presto-hive/src/main/java/io/prestosql/plugin/hive/MsckProcedure.java

```diff
@@ -0,0 +1,213 @@
+                        new Argument("schema_name", VARCHAR),
+                        new Argument("table_name", VARCHAR),
+                        new Argument("add", BOOLEAN),
+                        new Argument("drop", BOOLEAN)),
+                MSCK.bindTo(this));
+    }
+
+    public void msck(ConnectorSession session, String schemaName, String tableName, Boolean add, Boolean drop)
+    {
+        HdfsEnvironment.HdfsContext context = new HdfsEnvironment.HdfsContext(session, schemaName, tableName);
+        SemiTransactionalHiveMetastore metastore = ((HiveMetadata) hiveMetadataFactory.get()).getMetastore();
+
+        Table table = metastore.getTable(schemaName, tableName).orElse(null);
+        if (table == null) {
+            return;
+        }
+
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation);
+
+            List<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName).orElse(emptyList());
+            List<String> partitionsInFileSystem = listDirectory(fileSystem, fileSystem.getFileStatus(defaultLocation), table.getPartitionColumns(), table.getPartitionColumns().size()).stream()
+                    .map(fileStatus -> defaultLocation.toUri().relativize(fileStatus.getPath().toUri()).getPath())
+                    .collect(toList());
+
+            // partitions in metastore but not in file system
+            List<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            // partitions in file system but not in metastore
+            List<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+
+            if (add) {
+                addPartitions(metastore, session, table, defaultLocation, partitionsToAdd);
+            }
+            if (drop) {
+                dropPartitions(metastore, session, table, partitionsToDrop);
+            }
+            metastore.commit();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private List<FileStatus> listDirectory(FileSystem fileSystem, FileStatus current, List<Column> partitionColumns, int depth)
+    {
+        if (depth == 0) {
+            return ImmutableList.of(current);
+        }
+
+        try {
+            return Stream.of(fileSystem.listStatus(current.getPath()))
+                    .filter(fileStatus -> validatePath(fileSystem, fileStatus, partitionColumns.get(partitionColumns.size() - depth)))
+                    .flatMap(directory -> listDirectory(fileSystem, directory, partitionColumns, depth - 1).stream())
+                    .collect(toList());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean validatePath(FileSystem fileSystem, FileStatus file, Column column)
```

Comment:
Can i have paths like
`a=1/b=zoo`
`b=zoo/a=2`
?

I vaguely recall some tools can do this to you.

At the same time, we silently ignore any directories which we don't recognize.
It would be safer to reject disallow drop (throw) if there were subdirectories we didn't recognize. And have yet another parameter which needs to be set in order to ignore unrecognized subdirectories.

Also, if a subdirectory (subpath) is not a recognized partition, _but_ is listed on `partitionsInMetastore`, we should not drop it.

Reply:
> Can i have paths like
> `a=1/b=zoo`
> `b=zoo/a=2`
> ?
> 
> I vaguely recall some tools can do this to you.

I believe the partition columns are ordered. So if the table is defined as `partitioned_by = ARRAY[ 'a', 'b' ]` then only `a=1/b=zoo` is a valid partition directory.
@electrum pointed it out in the slack channel, and I checked the code it also seems to acknowledge the ordering. @dain's implementation of `FileHiveMetastore::getPartitionNames` list directories that follow the pre-defined order.

>At the same time, we silently ignore any directories which we don't recognize.
It would be safer to reject disallow drop (throw) if there were subdirectories we didn't recognize. And have yet another parameter which needs to be set in order to ignore unrecognized subdirectories.

This complicates the semantics of this operation. Also ignore unrecognized directories would save some time in the `listDirectory` by terminating the search, especially in cloud storage (object store). 

>Also, if a subdirectory (subpath) is not a recognized partition, _but_ is listed on `partitionsInMetastore`, we should not drop it.
Will that happen to a managed table?